elfi/distributions.py

```diff
@@ -0,0 +1,334 @@
+"""This module implements empirical estimations of distributions.
+
+Currently it is mainly used for Gaussian copula ABC.
+
+References
+----------
+Jingjing Li, David J. Nott, Yanan Fan, Scott A. Sisson
+Extending approximate Bayesian computation methods to high dimensions via Gaussian copula.
+Computational Statistics & Data Analysis
+Volume 106, February 2017, Pages 77-89
+https://doi.org/10.1016/j.csda.2016.07.005
+
+"""
+import numpy as np
+import scipy.stats as ss
+from scipy.interpolate import interp1d
+
+__all__ = ('EmpiricalDensity', 'ecdf', 'eppf', 'estimate_densities', 'MetaGaussian')
+
+
+def ecdf(samples):
+    """Compute an empirical cdf.
+
+    Parameters
+    ----------
+    samples : array_like
+      a univariate sample
+
+    Returns
+    -------
+    empirical_cdf
+      an interpolated function for the estimated cdf
+
+    """
+    x, y = _handle_endpoints(*_ecdf(samples))
+    return _interp_ecdf(x, y)
+
+
+def eppf(samples):
+    """Compute an empirical quantile function.
+
+    Parameters
+    ----------
+    samples : array_like
+      a univariate sample
+
+    Returns
+    -------
+    empirical_ppf
+      an interpolated function for the estimated quantile function
+
+    """
+    x, y = _handle_endpoints(*_ecdf(samples))
+    return _interp_ppf(x, y)
+
+
+def _ecdf(x):
+    xs = np.sort(x)
+    ys = np.arange(1, len(xs) + 1)/float(len(xs))
+    return xs, ys
+
+
+def _low(xs, ys):
+    """Compute the intercetion point (x, 0)."""
+    slope = (ys[1] - ys[0])/(xs[1] - xs[0])
+    intersect = -ys[0]/slope + xs[0]
+    return intersect
+
+
+def _high(xs, ys):
+    """Compute the interception point (x, 1)."""
+    slope = (ys[-1] - ys[-2])/(xs[-1] - xs[-2])
+    intersect = (1 - ys[-1])/slope + xs[-1]
+    return intersect
+
+
+def _handle_endpoints(xs, ys):
+    high = _high(xs, ys)
+    low = _low(xs, ys)
+
+    # add endpoints
+    x = np.append(np.insert(xs, 0, low), high)
+    y = np.append(np.insert(ys, 0, 0.), 1.)
+    return x, y
+
+
+def _interp_ecdf(x, y, **kwargs):
+    low, high = x[0], x[-1]
+
+    # linear interpolation
+    f = interp1d(x, y, **kwargs)
+
+    def interp(q):
+        if isinstance(q, np.ndarray):
+            too_low = sum(q < low)
+            too_high = sum(q > high)
+            return np.concatenate([np.zeros(too_low),
+                                   f(q[(q >= low) & (q <= high)]),
+                                   np.ones(too_high)])
+        else:
+            return _scalar_cdf(f, low, high, q)
+
+    return interp
+
+
+def _scalar_cdf(f, low, high, q):
+    if q < low:
+        return 0
+    elif q > high:
+        return 1
+    else:
+        return f(q)
+
+
+def _interp_ppf(x, y, **kwargs):
+    f = interp1d(y, x, **kwargs)
+
+    def interp(p):
+        try:
+            return f(p)
+        except ValueError:
+            raise ValueError("The quantile function is not defined outside [0, 1].")
+
+    return interp
+
+
+class EmpiricalDensity(object):
```

Comment:
Inherit and implement elfi.Distribution

Reply:
Not really possible to do that. It is not possible to implement the methods as classmethods because they depend on the data in the instance. I'l make the interface compatible so that they can be duck typed, which seems to be the more idiomatic way anyway.

Reply:
Hmm... Yes, I see the difficulty, but I'm sure it can be done somehow (possible requires changing elfi.Distribution). Otherwise there's very limited use for these in ELFI?

Reply:
Currently `elfi.Distribution` (`elfi.model.extensions.ScipyLikeDistribution`) doesn't really have any functionality. It is merely defining the interface of a scipy like distribution.  I would argue that this is better handled with duck typing instead of inheriting. The classes I implemented have the same interface as `elfi.Distribution` so they should be usable as is.

Until there is  functionality added to `elfi.Distribution`, I don't see the reason to inherit from it just for the sake of it. Even then,  the implementation of  of `elfi.Distribution` would have to be refactored to make it an actual 'is-a' relation. In fact, I think reuse through composition would be better in that case.

Reply:
Sorry, I thought that we check the type. But it seems we don't, and maybe it's not needed either. Ok.